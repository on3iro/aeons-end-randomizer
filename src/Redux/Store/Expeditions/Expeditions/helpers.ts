import shortid from 'shortid'

import * as types from 'types'
import { byCost } from 'helpers'

import {
  createArrayWithDefaultValues,
  createIdList,
  getRandomEntity,
  createSupply,
} from 'Redux/helpers'

import { BaseConfig, RollBattleConfig, WinConfig, LossConfig } from './types'

export const calcBattleScore = (tries: number) => {
  switch (tries) {
    case 1: {
      return 6
    }

    case 2: {
      return 4
    }

    case 3: {
      return 2
    }

    default: {
      return 0
    }
  }
}

export const determineBaseTier = () => {}

export const generateBattles = (
  variant: types.Variant,
  expeditionId: string
) => {
  const battles = variant.configList.map(
    (config, index): types.Battle => {
      const isFirst = index === 0

      return {
        id: shortid.generate(),
        nemesisTier: config.tier,
        treasure: config.treasure,
        expeditionId,
        status: isFirst ? 'unlocked' : 'locked',
        tries: 0,
      }
    }
  )

  return battles
}

export const createExpeditionConfig = ({
  variant,
  name,
  bigPocketVariant,
  availableMageIds,
  availableCards,
  availableLevel1TreasureIds,
  tiles,
}: BaseConfig) => {
  const mageIds = createIdList(
    availableMageIds,
    createArrayWithDefaultValues(4, 'EMPTY'),
    getRandomEntity
  ).result
  const { gems, relics, spells } = createSupply(availableCards, tiles)
  const gemsByCost = gems.result.sort(byCost)
  const relicsByCost = relics.result.sort(byCost)
  const spellsByCost = spells.result.sort(byCost)

  const supplyIds = [...gemsByCost, ...relicsByCost, ...spellsByCost].map(
    card => card.id
  )

  const startsWithTreasure = variant.configList[0].tier.tier > 1 // TODO should probably use our rollTreasure property
  const treasureIds = startsWithTreasure
    ? createIdList(
        availableLevel1TreasureIds,
        createArrayWithDefaultValues(5, 'EMPTY'),
        getRandomEntity
      ).result
    : []

  const expeditionId = shortid.generate()
  const battles = generateBattles(variant, expeditionId)

  return {
    battles,
    bigPocketVariant,
    expeditionId,
    mageIds,
    name,
    supplyIds,
    treasureIds,
    variantId: variant.id,
  }
}

// Because we always add these cards to an existing array inside our expedition in our store,
// the count of newly added cards decreases by tier.
// If we ever have a variant, that starts higher than tier 2 we have to
// incorporate changes to roll the whole amount of cards per tier!
const getUpgradedBasicNemesisIdsByBattleTier = ({
  availableTier1Ids,
  availableTier2Ids,
  availableTier3Ids,
  battleTier,
}: {
  availableTier1Ids: string[]
  availableTier2Ids: string[]
  availableTier3Ids: string[]
  battleTier: 1 | 2 | 3 | 4
}) => {
  switch (battleTier) {
    case 1: {
      return [] // No upgraded cards are added on tier 1
    }

    case 2: {
      const tier1Ids = createIdList(
        availableTier1Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier2Ids = createIdList(
        availableTier2Ids,
        createArrayWithDefaultValues(3, 'EMPTY'),
        getRandomEntity
      ).result
      const tier3Ids = createIdList(
        availableTier3Ids,
        createArrayWithDefaultValues(3, 'EMPTY'),
        getRandomEntity
      ).result

      return [...tier1Ids, ...tier2Ids, ...tier3Ids]
    }

    case 3:
    case 4: {
      const tier1Ids = createIdList(
        availableTier1Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier2Ids = createIdList(
        availableTier2Ids,
        createArrayWithDefaultValues(1, 'EMPTY'),
        getRandomEntity
      ).result
      const tier3Ids = createIdList(
        availableTier3Ids,
        createArrayWithDefaultValues(2, 'EMPTY'),
        getRandomEntity
      ).result

      return [...tier1Ids, ...tier2Ids, ...tier3Ids]
    }

    default: {
      return [] // Should never occur!
    }
  }
}

const rollNewUpgradedNemesisCards = (
  availableUpgradedBasicNemesisCards: types.UpgradedBasicNemesisCard[],
  previousUpgradedBasicNemesisCards: string[],
  nemesisTier: 1 | 2 | 3 | 4
) => {
  const upgradedCardsWithoutPreviousCards = availableUpgradedBasicNemesisCards.filter(
    upgradedCard => !previousUpgradedBasicNemesisCards.includes(upgradedCard.id)
  )

  const tier1AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 1)
    .map(card => card.id)
  const tier2AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 2)
    .map(card => card.id)
  const tier3AvailableUpgradedNemesisIds = upgradedCardsWithoutPreviousCards
    .filter(card => card.tier === 3)
    .map(card => card.id)

  const upgradedBasicNemesisCardIds = getUpgradedBasicNemesisIdsByBattleTier({
    battleTier: nemesisTier,
    availableTier1Ids: tier1AvailableUpgradedNemesisIds,
    availableTier2Ids: tier2AvailableUpgradedNemesisIds,
    availableTier3Ids: tier3AvailableUpgradedNemesisIds,
  })

  return [...previousUpgradedBasicNemesisCards, ...upgradedBasicNemesisCardIds]
}

export const createBattle = (config: RollBattleConfig) => {
  const nemesisIds = config.availableNemeses
    .filter(
      nemesis => nemesis.expeditionRating === config.battle.nemesisTier.tier
    )
    .map(nemesis => nemesis.id)
    .filter(nemesisId => !config.previousNemeses.includes(nemesisId))

  const nemesisId = createIdList(
    nemesisIds,
    createArrayWithDefaultValues(1, 'EMPTY'),
    getRandomEntity
  ).result[0]

  const upgradedBasicNemesisCardIds = config.battle.nemesisTier.isNewTier
    ? rollNewUpgradedNemesisCards(
        config.availableUpgradedBasicNemesisCards,
        config.previousUpgradedBasicNemesisCards,
        config.battle.nemesisTier.tier
      )
    : config.previousUpgradedBasicNemesisCards

  return {
    battle: { ...config.battle, nemesisId, status: 'before_battle' },
    upgradedBasicNemesisCardIds,
  }
}

const rollNewEntity = (list: string[]): string =>
  createIdList(list, createArrayWithDefaultValues(1, 'EMPTY'), getRandomEntity)
    .result[0]

export const rollWinRewards = (config: WinConfig) => {
  const newTreasures = createIdList(
    config.treasureIds,
    createArrayWithDefaultValues(5, 'EMPTY'),
    getRandomEntity
  ).result

  const newGem = rollNewEntity(config.gemIds)
  const newRelic = rollNewEntity(config.relicIds)
  const newSpell = rollNewEntity(config.spellIds)

  const updatedBattle = {
    ...config.battle,
    rewards: {
      treasure: newTreasures,
      supplyIds: [newGem, newRelic, newSpell],
      mage: undefined, // we explicitely overwrite the mage reward from earlier losses
    },
  }
  return { battle: updatedBattle }
}

export const rollLossRewards = (config: LossConfig) => {
  switch (config.type) {
    case 'mage': {
      const newMage = rollNewEntity(config.mageIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: newMage,
          supplyIds: [],
        },
      }
    }

    case 'gem': {
      const newGem = rollNewEntity(config.gemIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newGem],
        },
      }
    }

    case 'relic': {
      const newRelic = rollNewEntity(config.relicIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newRelic],
        },
      }
    }

    case 'spell': {
      const newSpell = rollNewEntity(config.spellIds)

      return {
        ...config.battle,
        rewards: {
          treasure: [],
          mage: undefined,
          supplyIds: [newSpell],
        },
      }
    }

    case 'treasure1': {
      const newTreasure = rollNewEntity(config.treasure1Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    case 'treasure2': {
      const newTreasure = rollNewEntity(config.treasure2Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    case 'treasure3': {
      const newTreasure = rollNewEntity(config.treasure3Ids)

      return {
        ...config.battle,
        rewards: {
          treasure: [newTreasure],
          mage: undefined,
          supplyIds: [],
        },
      }
    }

    default: {
      return { ...config.battle }
    }
  }
}
