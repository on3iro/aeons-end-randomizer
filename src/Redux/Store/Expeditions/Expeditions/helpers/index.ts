export * from './convertExpeditionToConfig'
export * from './convertExpeditionFromConfig'

export const calcBattleScore = (tries: number) => {
  switch (tries) {
    case 1: {
      return 6
    }

    case 2: {
      return 4
    }

    case 3: {
      return 2
    }

    default: {
      return 0
    }
  }
}
