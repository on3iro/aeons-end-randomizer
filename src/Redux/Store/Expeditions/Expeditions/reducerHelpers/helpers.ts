import { loop, Cmd } from 'redux-loop'
import { set as setToDb } from 'idb-keyval'
import * as types from 'aer-types/types'

import { State } from '../types'
import { actions } from '../actions'

export const EXPEDITIONS_DB_KEY = 'expeditions-1.11.0'

export const updateBattle = (
  state: State,
  battle: types.Battle,
  additionalBattleProps?: {
    status?: types.BattleStatus
    tries?: number
    rewards?: types.Rewards
  }
) => {
  const oldExpedition = state.expeditions[battle.expeditionId]
  const branches = oldExpedition.sequence.branches

  const updatedBranches = {
    ...branches,
    [battle.id]: {
      ...battle,
      ...additionalBattleProps,
    },
  }

  const newState = {
    ...state,
    expeditions: {
      ...state.expeditions,
      [battle.expeditionId]: {
        ...oldExpedition,
        sequence: {
          ...oldExpedition.sequence,
          branches: updatedBranches,
        },
      },
    },
  }

  return loop(
    newState,
    Cmd.run(setToDb, {
      args: [EXPEDITIONS_DB_KEY, newState],
      successActionCreator: actions.setToDBSuccessful,
      failActionCreator: actions.setToDBFailed,
    })
  )
}
