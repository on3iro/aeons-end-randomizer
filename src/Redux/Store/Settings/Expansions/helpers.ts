import * as Languages from './Languages'

import {
  Nemesis,
  Mage,
  Treasure,
  BasicNemesisCard,
  UpgradedBasicNemesisCard,
  ICard,
} from 'aer-types/types'

import { LanguageKey } from './types'

type Entity =
  | Nemesis
  | Mage
  | Treasure
  | BasicNemesisCard
  | UpgradedBasicNemesisCard
  | ICard

export type ContentStruct<T> = {
  [key in LanguageKey]: {
    [id: string]: T
  }
}

export const getContentByIdWithLanguageFallback = <T extends Entity>(
  languages: Languages.State,
  content: ContentStruct<T>,
  id: string
) => {
  // Just get the corresponding expansion id from the english version
  const language = languages[content.ENG[id].expansion]

  // Fallback to english
  return content[language][id] || content.ENG[id]
}

export const getEntitiesByIdListWithLanguageFallback = <T extends Entity>(
  content: ContentStruct<T>,
  ids: string[],
  language: LanguageKey
) => {
  return ids.map((id) => {
    return content[language][id] || content.ENG[id]
  })
}
