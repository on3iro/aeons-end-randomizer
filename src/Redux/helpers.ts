import * as types from '../types'

type CardListReduceResult = {
  availableCards: types.ICard[]
  result: types.ICard[]
}

/**
 * Given a card object and a setupBlueprint this returns either
 * true or false, depending on the cost of the card and the expectation set by
 * the blueprints operation and cost.
 */
export const filterByCost = (
  card: { cost: number },
  setupBlueprint: {
    operation: types.Operation
    threshold?: number
    values?: Array<number>
  }
) => {
  switch (setupBlueprint.operation) {
    case '<': {
      if (!setupBlueprint.threshold) return true
      return card.cost < setupBlueprint.threshold
    }
    case '>': {
      if (!setupBlueprint.threshold) return true
      return card.cost > setupBlueprint.threshold
    }
    case '<=': {
      if (!setupBlueprint.threshold) return true
      return card.cost <= setupBlueprint.threshold
    }
    case '>=': {
      if (!setupBlueprint.threshold) return true
      return card.cost >= setupBlueprint.threshold
    }
    case '=': {
      if (!setupBlueprint.threshold) return true
      return card.cost === setupBlueprint.threshold
    }
    case 'OR': {
      if (!setupBlueprint.values) return true
      return setupBlueprint.values.indexOf(card.cost) !== -1
    }
    default: {
      return true
    }
  }
}

export const createCardList = (
  availableCards: types.ICard[],
  blueprints: Array<types.IBluePrint>,
  getEntity: <T>(list: Array<T>) => T
): CardListReduceResult =>
  blueprints.reduce(
    (acc: CardListReduceResult, blueprint: types.IBluePrint) => {
      const filteredByCost = acc.availableCards.filter(card =>
        filterByCost(card, blueprint)
      )

      // Skip blueprint if no compatible card is left
      if (filteredByCost.length <= 0) {
        return acc
      }

      const card = getEntity(filteredByCost)

      // Make sure each entity will only be added to the result list once
      const remainingCards = acc.availableCards.filter(
        entity => entity.id !== card.id
      )

      return {
        availableCards: remainingCards,
        result: [...acc.result, card],
      }
    },
    { availableCards, result: [] }
  )

/**
 * Creates a list of randomly selected cards from a given pool of cards by a specific type.
 * The amount of cards generated is determined by the amount of tileSetups matching
 * the given cardType. E.g. if 2 out of 9 tileSetups have the type="Gem" and the
 * specified type is "Gem" as well, a list of two cards will be created.
 *
 * return List<Card>
 */
const getRandomCardsByType = (
  availableCards: ReadonlyArray<types.ICard>,
  tileSetups: ReadonlyArray<types.IBluePrint>,
  cardType: types.CardType
): CardListReduceResult => {
  const cardSlots = tileSetups.filter(({ type }) => type === cardType)
  const availableCardsOfType = availableCards.filter(
    ({ type }) => type === cardType
  )
  return createCardList(availableCardsOfType, cardSlots, getRandomEntity)
}

/**
 * Creates a supply object, containing gems, relics and spells with meta information.
 * The pool of available cards will be determined by the sets the user has selected
 * inside the application settings. The supply will be created via the blueprint
 * of the given tileSetups.
 *
 * return { gems, relics, spells }
 */
export const createSupply = (
  availableCards: ReadonlyArray<types.ICard>,
  tileSetups: ReadonlyArray<types.IBluePrint>
) => {
  const gems = getRandomCardsByType(availableCards, tileSetups, 'Gem')
  const relics = getRandomCardsByType(availableCards, tileSetups, 'Relic')
  const spells = getRandomCardsByType(availableCards, tileSetups, 'Spell')

  return { gems, relics, spells }
}

export const createArrayWithDefaultValues = (
  size: number,
  defaultValue: any
) => {
  return Array.from({ length: size }, () => defaultValue)
}

export const createSlotList = (amount: number): Array<types.IEmptyBluePrint> =>
  createArrayWithDefaultValues(amount, { type: 'EMPTY', operation: 'NoOp' })

type Entity = { id: string } | string

export const generateListFrom = <T extends Entity>(
  availableEntities: Array<T> | ReadonlyArray<T>,
  slots: Array<any>,
  getEntity: <E extends Entity>(list: Array<E>) => E
): { availableEntities: Array<T>; result: Array<T> } => {
  const result = slots.reduce(
    (acc, _) => {
      // If no entity is left, simply return the actual empty slot
      const entity = getEntity(acc.availableEntities)

      if (!entity) {
        return acc
      }

      // Make sure each entity will only be added to the result list once
      const remainingEntities = acc.availableEntities.filter(
        (available: Entity) => {
          if (typeof available === 'string' && typeof entity === 'string') {
            return available !== entity
          }

          if (typeof available === 'object' && typeof entity === 'object') {
            return available.id !== entity.id
          }

          return true
        }
      )

      return {
        availableEntities: remainingEntities,
        result: [...acc.result, entity],
      }
    },
    { availableEntities, result: [] }
  )

  return result
}

export const createTurnOrderCardList = (
  availableCards: types.ITurnOrderCard[],
  slots: types.ITurnOrderCard[],
  getEntity: <T>(list: Array<T>) => T
) => generateListFrom(availableCards, slots, getEntity)

export const createMageList = (
  availableMages: ReadonlyArray<types.Mage>,
  slots: Array<types.Slot>,
  getEntity: <T>(list: Array<T>) => T
) => generateListFrom(availableMages, slots, getEntity)

export const createIdList = (
  availableIds: ReadonlyArray<string>,
  slots: string[],
  getEntity: <T>(list: Array<T>) => T
) => generateListFrom(availableIds, slots, getEntity)

export const shuffleDeck = (
  deck: types.ITurnOrderCard[]
): types.ITurnOrderCard[] => {
  return createTurnOrderCardList(deck, deck, getRandomEntity).result
}

/**
 * Gets a random value from a list. (The wording of entities is just used for semantic context)
 */
export const getRandomEntity = <E>(availableEntities: ReadonlyArray<E>) =>
  availableEntities[Math.floor(Math.random() * availableEntities.length)]

export const getOperationString = (
  operation: types.Operation,
  values?: number[],
  threshold?: number
) => {
  if (operation === 'NoOp') {
    return ''
  }

  if (operation === 'ANY') {
    return operation
  }

  if (operation === 'OR' && values) {
    return values.join('/')
  }

  const thresholdValue = threshold ? threshold : ''

  return `${operation} ${thresholdValue}`
}
