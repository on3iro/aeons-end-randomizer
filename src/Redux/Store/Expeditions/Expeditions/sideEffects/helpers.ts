import * as types from 'aer-types/types'

import {
  createArrayWithDefaultValues,
  createIdList,
  getRandomEntity,
  createCardList,
} from 'Redux/helpers'

import { selectors } from 'Redux/Store'
import { ExpeditionsStateSlice } from '../types'
import { TreasureIdsStateSlice } from 'Redux/Store/Settings/Expansions/Treasures/ids'
import { TreasureContentStateSlice } from 'Redux/Store/Settings/Expansions/Treasures/content'
import { MagesContentStateSlice } from 'Redux/Store/Settings/Expansions/Mages/content'
import { CardsContentStateSlice } from 'Redux/Store/Settings/Expansions/Cards/content'
import { SelectedLanguagesStateSlice } from 'Redux/Store/Settings/Expansions/Languages'

export const rollNewEntity = (
  list: string[],
  getEntity: types.SeededEntityGetter = getRandomEntity,
  seed: types.Seed
): { result: string; seed: types.Seed } => {
  const idList = createIdList(
    list,
    createArrayWithDefaultValues(1, 'EMPTY'),
    getEntity,
    seed
  )
  return {
    result: idList.result[0],
    seed: idList.seed,
  }
}

type RewardResult = {
  result: string[]
  seed: types.Seed
}

// TODO add tests
export const getSupplyIds = ({
  supply,
  seed,
  stillAvailableCardsByType,
}: {
  supply: types.SupplyRewardConfig | undefined
  seed: types.Seed
  stillAvailableCardsByType: {
    Gem: types.ICard[]
    Relic: types.ICard[]
    Spell: types.ICard[]
    EMPTY: never[]
    ANY: never[]
  }
}) => {
  const baseResult = {
    result: [],
    seed,
  }

  if (supply === undefined) {
    return baseResult
  } else {
    // We split blueprints and ids.
    // This has two benefits:
    // 1.) it lets us filter available cards by those ids, which have been explicitely specified
    // 2.) it makes the reduce call a bit easer to read and maintain, as we don't have to handle the 'string' case
    const ids = supply.ids.filter(
      (idOrBlueprint) => typeof idOrBlueprint === 'string'
    )
    const bluePrints = supply.ids.filter(
      (idOrBlueprint): idOrBlueprint is types.IBluePrint =>
        typeof idOrBlueprint !== 'string'
    )

    return bluePrints.reduce(
      (acc: RewardResult, blueprint) => {
        const cardType = blueprint.type
        const stillAvailableCards = stillAvailableCardsByType[cardType].filter(
          (card) => ids.indexOf(card.id) === -1
        )

        const cardCreationResult = createCardList(
          stillAvailableCards,
          [blueprint],
          undefined,
          acc.seed
        )

        return {
          result: [
            ...acc.result,
            ...cardCreationResult.result.map((card) => card.id),
          ],
          seed: cardCreationResult.seed,
        }
      },
      {
        ...baseResult,
        result: ids,
      } as RewardResult // FIXME casting should technically not be necessary (but i could make it work without)
    )
  }
}

export const getTreasureIds = ({
  treasure,
  stillAvailableTreasureIdsByLevel,
  seed,
}: {
  treasure: types.TreasureRewardConfig | undefined
  seed: types.Seed
  stillAvailableTreasureIdsByLevel: {
    1: string[]
    2: string[]
    3: string[]
  }
}) => {
  const baseResult = {
    result: [],
    seed,
  }

  if (treasure === undefined) {
    return baseResult
  } else {
    const ids = treasure.ids.filter(
      (idOrRandom) => typeof idOrRandom === 'string'
    )

    const randomTreasureConfigs = treasure.ids.filter(
      (idOrRandom): idOrRandom is { random: true; level: 1 | 2 | 3 } =>
        typeof idOrRandom !== 'string'
    )

    return randomTreasureConfigs.reduce(
      (acc: RewardResult, config) => {
        const treasureLevel = config.level
        const stillAvailableCards = stillAvailableTreasureIdsByLevel[
          treasureLevel
        ].filter((treasureId) => ids.indexOf(treasureId) === -1)

        const treasureIdsResult = createIdList(
          stillAvailableCards,
          ['EMPTY'],
          undefined,
          acc.seed
        )

        return {
          result: [...acc.result, ...treasureIdsResult.result],
          seed: treasureIdsResult.seed,
        }
      },
      {
        ...baseResult,
        result: ids,
      } as RewardResult // FIXME casting should technically not be necessary (but i could make it work without)
    )
  }
}

export const getMageIds = ({
  mage,
  stillAvailableMageIds,
  seed,
}: {
  mage: types.MageRewardConfig | undefined
  seed: types.Seed
  stillAvailableMageIds: string[]
}) => {
  const baseResult = {
    result: [],
    seed,
  }

  if (mage === undefined) {
    return baseResult
  } else {
    const ids = mage.ids.filter((idOrRandom) => typeof idOrRandom === 'string')

    const randomTreasureConfigs = mage.ids.filter(
      (idOrRandom): idOrRandom is { random: true } =>
        typeof idOrRandom !== 'string'
    )

    return randomTreasureConfigs.reduce(
      (acc: RewardResult, _) => {
        const filteredMageIds = stillAvailableMageIds.filter(
          (mageId) => ids.indexOf(mageId) === -1
        )

        const mageIdsResult = createIdList(
          filteredMageIds,
          ['EMPTY'],
          undefined,
          acc.seed
        )

        return {
          result: [...acc.result, ...mageIdsResult.result],
          seed: mageIdsResult.seed,
        }
      },
      {
        ...baseResult,
        result: ids,
      } as RewardResult // FIXME casting should technically not be necessary (but i could make it work without)
    )
  }
}

// TODO add tests
export const handleCustomRewards = (
  state: ExpeditionsStateSlice &
    CardsContentStateSlice &
    TreasureContentStateSlice &
    TreasureIdsStateSlice &
    MagesContentStateSlice &
    SelectedLanguagesStateSlice,
  rewardConfig: {
    type: 'custom'
    treasure?: types.TreasureRewardConfig
    mage?: types.MageRewardConfig
    supply?: types.SupplyRewardConfig
  },
  expedition: types.Expedition
) => {
  const expeditionId = expedition.id

  ///////////////////////
  // Content Selection //
  ///////////////////////

  const stillAvailableCardsByType = {
    Gem: selectors.getStillAvailableGems(state, {
      expeditionId,
    }),
    Relic: selectors.getStillAvailableRelics(state, {
      expeditionId,
    }),
    Spell: selectors.getStillAvailableSpells(state, {
      expeditionId,
    }),
    EMPTY: [],
    ANY: [],
  }

  const stillAvailableTreasureIdsByLevel = {
    1: selectors.getStillAvailableTreasureIdsByLevel(state, {
      treasureLevel: 1,
      expeditionId,
    }),
    2: selectors.getStillAvailableTreasureIdsByLevel(state, {
      treasureLevel: 2,
      expeditionId,
    }),
    3: selectors.getStillAvailableTreasureIdsByLevel(state, {
      treasureLevel: 3,
      expeditionId,
    }),
  }

  const stillAvailableMageIds = selectors.getStillAvailableMageIds(state, {
    expeditionId,
  })

  ////////////
  // Supply //
  ////////////

  const { supply } = rewardConfig
  const supplyIdsResult = getSupplyIds({
    supply,
    seed: {
      seed: expedition.seed.seed,
      state: expedition.seed.supplyState,
    },
    stillAvailableCardsByType,
  })

  ///////////
  // Mages //
  ///////////

  const { mage } = rewardConfig
  const mageIdsResult = getMageIds({
    mage,
    seed: supplyIdsResult.seed,
    stillAvailableMageIds,
  })

  ///////////////
  // Treasures //
  ///////////////

  const { treasure } = rewardConfig
  const treasureIdsResult = getTreasureIds({
    treasure,
    seed: mageIdsResult.seed,
    stillAvailableTreasureIdsByLevel,
  })

  return {
    rewards: {
      supplyIds: supplyIdsResult.result,
      treasure: treasureIdsResult.result,
      mages: mageIdsResult.result,
    },
    seed: treasureIdsResult.seed,
  }
}
