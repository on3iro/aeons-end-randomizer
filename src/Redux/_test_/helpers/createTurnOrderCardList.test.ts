import { createTurnOrderCardList } from '../../helpers'
import { TURNORDERSETUPS } from 'config/turnOrderSetups'

describe('createTurnOrderCardList()', () => {
  it('should return the correct elements with getFirst() entity getter', () => {
    const getFirst = (available: any[]) => available[0]

    const result = createTurnOrderCardList(
      TURNORDERSETUPS.onePlayer.variations.default.turnOrderCards,
      TURNORDERSETUPS.onePlayer.variations.default.turnOrderCards,
      getFirst
    )

    expect(result).toEqual({
      availableEntities: [],
      result: TURNORDERSETUPS.onePlayer.variations.default.turnOrderCards,
    })
  })

  it('should return the correct elements with getSecond() entity getter', () => {
    const getSecond = (available: any[]) => available[1]

    const result = createTurnOrderCardList(
      TURNORDERSETUPS.onePlayer.variations.default.turnOrderCards,
      TURNORDERSETUPS.onePlayer.variations.default.turnOrderCards,
      getSecond
    )

    expect(result).toEqual({
      availableEntities: [
        {
          id: 'player1-1',
          name: 'Player 1',
          type: 'player1',
        },
      ],
      result: [
        { id: 'player1-2', name: 'Player 1', type: 'player1' },
        { id: 'player1-3', name: 'Player 1', type: 'player1' },
        { id: 'nemesis-1', name: 'Nemesis', type: 'nemesis' },
        { id: 'nemesis-2', name: 'Nemesis', type: 'nemesis' },
      ],
    })
  })
})
