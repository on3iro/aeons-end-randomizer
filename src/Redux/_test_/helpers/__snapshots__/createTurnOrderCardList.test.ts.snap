// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTurnOrderCardList() test 1 1`] = `
Object {
  "availableEntities": Array [],
  "result": Array [
    Object {
      "id": "nemesis-2",
      "name": "Nemesis",
      "type": "nemesis",
    },
    Object {
      "id": "player1-2",
      "name": "Player 1",
      "type": "player1",
    },
    Object {
      "id": "nemesis-1",
      "name": "Nemesis",
      "type": "nemesis",
    },
    Object {
      "id": "player1-1",
      "name": "Player 1",
      "type": "player1",
    },
    Object {
      "id": "player1-3",
      "name": "Player 1",
      "type": "player1",
    },
  ],
  "seed": Object {
    "seed": "test",
    "state": Object {
      "S": Array [
        92,
        55,
        235,
        213,
        126,
        59,
        37,
        65,
        117,
        41,
        205,
        7,
        51,
        174,
        214,
        0,
        171,
        155,
        139,
        1,
        196,
        125,
        138,
        225,
        223,
        97,
        147,
        203,
        109,
        102,
        190,
        246,
        173,
        81,
        152,
        250,
        79,
        192,
        53,
        154,
        49,
        218,
        164,
        40,
        35,
        224,
        101,
        242,
        186,
        12,
        248,
        200,
        252,
        142,
        160,
        111,
        249,
        17,
        91,
        88,
        130,
        232,
        96,
        140,
        66,
        54,
        69,
        76,
        75,
        222,
        188,
        149,
        34,
        177,
        184,
        191,
        9,
        178,
        238,
        68,
        217,
        187,
        131,
        189,
        185,
        234,
        194,
        143,
        28,
        121,
        167,
        150,
        100,
        255,
        64,
        106,
        86,
        38,
        204,
        123,
        215,
        170,
        114,
        87,
        195,
        115,
        183,
        19,
        31,
        95,
        193,
        62,
        85,
        105,
        60,
        241,
        180,
        156,
        159,
        208,
        210,
        15,
        244,
        89,
        163,
        108,
        237,
        116,
        3,
        13,
        30,
        129,
        136,
        176,
        112,
        70,
        46,
        61,
        239,
        198,
        165,
        110,
        209,
        230,
        94,
        221,
        107,
        228,
        5,
        207,
        226,
        206,
        63,
        50,
        20,
        52,
        122,
        201,
        145,
        157,
        48,
        148,
        219,
        236,
        181,
        104,
        128,
        254,
        135,
        2,
        43,
        82,
        16,
        166,
        212,
        18,
        134,
        146,
        29,
        168,
        10,
        83,
        151,
        78,
        119,
        93,
        67,
        77,
        133,
        227,
        118,
        197,
        132,
        45,
        175,
        216,
        153,
        26,
        211,
        161,
        144,
        58,
        243,
        23,
        245,
        4,
        74,
        27,
        251,
        24,
        253,
        11,
        80,
        14,
        21,
        162,
        137,
        231,
        240,
        57,
        179,
        71,
        33,
        120,
        124,
        8,
        98,
        32,
        72,
        169,
        247,
        22,
        229,
        25,
        127,
        42,
        233,
        99,
        36,
        73,
        172,
        141,
        182,
        6,
        39,
        56,
        202,
        47,
        90,
        199,
        158,
        220,
        103,
        113,
        44,
        84,
      ],
      "i": 35,
      "j": 166,
    },
  },
}
`;

exports[`createTurnOrderCardList() test 2 1`] = `
Object {
  "availableEntities": Array [],
  "result": Array [
    Object {
      "id": "nemesis-2",
      "name": "Nemesis",
      "type": "nemesis",
    },
    Object {
      "id": "player1-2",
      "name": "Player 1",
      "type": "player1",
    },
    Object {
      "id": "nemesis-1",
      "name": "Nemesis",
      "type": "nemesis",
    },
    Object {
      "id": "player1-1",
      "name": "Player 1",
      "type": "player1",
    },
    Object {
      "id": "player1-3",
      "name": "Player 1",
      "type": "player1",
    },
  ],
  "seed": Object {
    "seed": "test",
    "state": Object {
      "S": Array [
        92,
        55,
        235,
        213,
        126,
        59,
        37,
        65,
        117,
        41,
        205,
        7,
        51,
        174,
        214,
        0,
        171,
        155,
        139,
        1,
        196,
        125,
        138,
        225,
        223,
        97,
        147,
        203,
        109,
        102,
        190,
        246,
        173,
        81,
        152,
        250,
        79,
        192,
        53,
        154,
        49,
        218,
        164,
        40,
        35,
        224,
        101,
        242,
        186,
        12,
        248,
        200,
        252,
        142,
        160,
        111,
        249,
        17,
        91,
        88,
        130,
        232,
        96,
        140,
        66,
        54,
        69,
        76,
        75,
        222,
        188,
        149,
        34,
        177,
        184,
        191,
        9,
        178,
        238,
        68,
        217,
        187,
        131,
        189,
        185,
        234,
        194,
        143,
        28,
        121,
        167,
        150,
        100,
        255,
        64,
        106,
        86,
        38,
        204,
        123,
        215,
        170,
        114,
        87,
        195,
        115,
        183,
        19,
        31,
        95,
        193,
        62,
        85,
        105,
        60,
        241,
        180,
        156,
        159,
        208,
        210,
        15,
        244,
        89,
        163,
        108,
        237,
        116,
        3,
        13,
        30,
        129,
        136,
        176,
        112,
        70,
        46,
        61,
        239,
        198,
        165,
        110,
        209,
        230,
        94,
        221,
        107,
        228,
        5,
        207,
        226,
        206,
        63,
        50,
        20,
        52,
        122,
        201,
        145,
        157,
        48,
        148,
        219,
        236,
        181,
        104,
        128,
        254,
        135,
        2,
        43,
        82,
        16,
        166,
        212,
        18,
        134,
        146,
        29,
        168,
        10,
        83,
        151,
        78,
        119,
        93,
        67,
        77,
        133,
        227,
        118,
        197,
        132,
        45,
        175,
        216,
        153,
        26,
        211,
        161,
        144,
        58,
        243,
        23,
        245,
        4,
        74,
        27,
        251,
        24,
        253,
        11,
        80,
        14,
        21,
        162,
        137,
        231,
        240,
        57,
        179,
        71,
        33,
        120,
        124,
        8,
        98,
        32,
        72,
        169,
        247,
        22,
        229,
        25,
        127,
        42,
        233,
        99,
        36,
        73,
        172,
        141,
        182,
        6,
        39,
        56,
        202,
        47,
        90,
        199,
        158,
        220,
        103,
        113,
        44,
        84,
      ],
      "i": 35,
      "j": 166,
    },
  },
}
`;
